tracing/src/main/java/com/palantir/remoting3/tracing/Tracers.java

```diff
@@ -177,19 +171,38 @@ public V call() throws Exception {
      */
     private static class TracingAwareRunnable implements Runnable {
         private final Runnable delegate;
-        private final Trace trace;
+        private DeferredTracer deferredTracer;
 
         TracingAwareRunnable(Runnable delegate) {
             this.delegate = delegate;
-            this.trace = Tracer.copyTrace();
+            this.deferredTracer = new DeferredTracer();
         }
 
         @Override
         public void run() {
+            deferredTracer.withTrace(() -> {
+                delegate.run();
+                return null;
+            });
+        }
+    }
+
+    public interface ThrowingCallable<T, E extends Throwable> {
```

Comment:
am I being silly/is there some existing interface like this?

Reply:
sadly nothing exists